File: tang/src/main/java/com/microsoft/tang/util/MonotonicSet.java
package com.microsoft.tang.util;

import java.util.Collection;
import java.util.Comparator;
import java.util.TreeSet;

public class MonotonicSet<T> extends TreeSet<T> {
  private static final long serialVersionUID = 1L;

  public MonotonicSet() {
    super();
  }

  public MonotonicSet(Collection<T> c) {
    addAll(c);
  }
  public MonotonicSet(Comparator<T> c) {
    super(c);
  }

  @Override
  public boolean add(T e) {
    if (super.contains(e)) {
      throw new IllegalArgumentException("Attempt to re-add " + e
          + " to MonotonicSet!");
    }
    return super.add(e);
  }

  @Override
  public void clear() {
    throw new UnsupportedOperationException("Attempt to clear MonotonicSet!");
  }

  @Override
  public boolean remove(Object o) {
    throw new UnsupportedOperationException("Attempt to remove " + o
        + " from MonotonicSet!");
  }

  @Override
  public boolean removeAll(Collection<?> c) {
    throw new UnsupportedOperationException(
        "removeAll() doesn't make sense for MonotonicSet!");
  }

  @Override
  public boolean retainAll(Collection<?> c) {
    throw new UnsupportedOperationException(
        "retainAll() doesn't make sense for MonotonicSet!");
  }

  @Override
  public boolean addAll(Collection<? extends T> c) {
    for (T t : c) {
      add(t);
    }
    return c.size() != 0;
  }

  public boolean addAllIgnoreDuplicates(Collection<? extends T> c) {
    boolean ret = false;
    for (T t : c) {
      if (!contains(t)) {
        add(t);
        ret = true;
      }
    }
    return ret;
  }
}

File: reef-tang/tang/src/main/cs/Tang/Util/MonotonicSet.cs
/**
 * Copyright (C) 2014 Microsoft Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
﻿using System;
using System.Collections.Generic;
using Com.Microsoft.Tang.Exceptions;

namespace Com.Microsoft.Tang.Util
{
    public class MonotonicSet<T> : SortedSet<T>
    {
        private static readonly long serialVersionUID = 1L;
        public MonotonicSet() : base()
        {
        }

        //public MonotonicSet(SortedSet<T> c) : base(c.Comparer)
        //{
        //    AddAll(c);
        //}

        public MonotonicSet(ICollection<T> c)
            : base(c)
        {
        }

        public MonotonicSet(IComparer<T> c)
            : base(c)
        {
        }

        public bool Add(T e)
        {
            if (this.Contains(e))
            {
                throw new ArgumentException("Attempt to re-add " + e
                    + " to MonotonicSet!");
            }
            return base.Add(e);
        }

      
        public bool AddAll(ICollection<T> c) 
        {
            foreach (T t in c)
            {
                this.Add(t);
            }

            return c.Count != 0;
        }

        
        public bool AddAllIgnoreDuplicates(ICollection<T> c) 
        {
            bool ret = false;
            foreach (T t in c) 
            {
                if (!Contains(t)) 
                {
                    Add(t);
                    ret = true;
                }
            }
            return ret;
        }

        public override void Clear() 
        {
            throw new UnsupportedOperationException("Attempt to clear MonotonicSet!");
        }

        public bool Remove(Object o)
        {
            throw new UnsupportedOperationException("Attempt to remove " + o
                + " from MonotonicSet!");
        }
    }
}
